substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1011 @@
+
+				ValidTransaction::with_tag_prefix("Sassafras")
+					.priority(TransactionPriority::max_value())
+					.longevity(tickets_longevity)
+					.and_provides(tickets_tag)
+					.propagate(true)
+					.build()
+			} else {
+				InvalidTransaction::Call.into()
+			}
+		}
+	}
+}
+
+// Inherent methods
+impl<T: Config> Pallet<T> {
+	/// Determine whether an epoch change should take place at this block.
+	/// Assumes that initialization has already taken place.
+	pub fn should_end_epoch(now: BlockNumberFor<T>) -> bool {
+		// The epoch has technically ended during the passage of time between this block and the
+		// last, but we have to "end" the epoch now, since there is no earlier possible block we
+		// could have done it.
+		//
+		// The exception is for block 1: the genesis has slot 0, so we treat epoch 0 as having
+		// started at the slot of block 1. We want to use the same randomness and validator set as
+		// signalled in the genesis, so we don't rotate the epoch.
+		now != One::one() && Self::current_slot_index() >= T::EpochDuration::get()
+	}
+
+	/// Current slot index with respect to current epoch.
+	fn current_slot_index() -> u64 {
+		Self::slot_index(CurrentSlot::<T>::get())
+	}
+
+	/// Slot index with respect to current epoch.
+	fn slot_index(slot: Slot) -> u64 {
+		if *GenesisSlot::<T>::get() == 0 {
+			return 0
+		}
+		slot.checked_sub(Self::current_epoch_start().into()).unwrap_or(u64::MAX)
+	}
+
+	/// Remove all tickets related data.
+	///
+	/// May not be efficient as the calling places may repeat some of this operations
+	/// but is a very extraordinary operation (hopefully never happens in production)
+	/// and better safe than sorry.
+	fn reset_tickets_data() {
+		let tickets_metadata = TicketsMeta::<T>::get();
+
+		// Remove even-epoch data.
+		let tickets_count = tickets_metadata.tickets_count[0];
+		(0..tickets_count).into_iter().for_each(|idx| {
+			if let Some(id) = TicketsIds::<T>::get((0, idx)) {
+				TicketsData::<T>::remove(id);
+			}
+		});
+
+		// Remove odd-epoch data.
+		let tickets_count = tickets_metadata.tickets_count[1];
+		(0..tickets_count).into_iter().for_each(|idx| {
+			if let Some(id) = TicketsIds::<T>::get((1, idx)) {
+				TicketsData::<T>::remove(id);
+			}
+		});
+
+		// Remove all outstanding tickets segments.
+		(0..tickets_metadata.segments_count).into_iter().for_each(|i| {
+			NextTicketsSegments::<T>::remove(i);
+		});
+		NextTicketsSegments::<T>::remove(u32::MAX);
+
+		// Reset tickets metadata
+		TicketsMeta::<T>::set(Default::default());
+	}
+
+	/// Enact an epoch change.
+	///
+	/// Should be done on every block where `should_end_epoch` has returned `true`, and the caller
+	/// is the only caller of this function.
+	///
+	/// Typically, this is not handled directly, but by a higher-level component implementing the
+	/// `EpochChangeTrigger` or `OneSessionHandler` trait.
+	///
+	/// If we detect one or more skipped epochs the policy is to use the authorities and values
```

Comment:
TODO: need to keep track of how many epochs were skipped in order to provide a correct mapping between epoch and session (see babe code)